ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/ConflictCheckingLedgerBridge.scala

```diff
@@ -0,0 +1,362 @@
+    @volatile var offsetIdx: Long = fromOffset(initialLedgerEnd)
+
+    in => {
+      Timed.value(
+        bridgeMetrics.Stages.sequence, {
+          offsetIdx = offsetIdx + 1
+          val newOffset = toOffset(offsetIdx)
+
+          val update = in match {
+            case Left(rejection) => rejection
+            case Right((_, NoOpPreparedSubmission(other))) =>
+              successMapper(other, offsetIdx, participantId)
+            case Right((noConflictUpTo, txSubmission: PreparedTransactionSubmission)) =>
+              val submitterInfo = txSubmission.originalSubmission.submitterInfo
+
+              withErrorLogger(submitterInfo.submissionId) { implicit errorLogger =>
+                conflictCheckWithInFlight(
+                  keysState = sequencerQueueState.keyState,
+                  consumedContractsState = sequencerQueueState.consumedContractsState,
+                  keyInputs = txSubmission.keyInputs,
+                  inputContracts = txSubmission.inputContracts,
+                )
+              }.fold(
+                toCommandRejectedUpdate(_, submitterInfo.toCompletionInfo),
+                { _ =>
+                  // Update the sequencer state
+                  sequencerQueueState = sequencerQueueState
+                    .dequeue(noConflictUpTo)
+                    .enqueue(newOffset, txSubmission.updatedKeys, txSubmission.consumedContracts)
+
+                  successMapper(txSubmission.originalSubmission, offsetIdx, participantId)
+                },
+              )
+          }
+
+          Iterable(newOffset -> update)
+        },
+      )
+    }
+  }
+
+  private def conflictCheckWithInFlight(
+      keysState: Map[Key, (Option[ContractId], LastUpdatedAt)],
+      consumedContractsState: Set[ContractId],
+      keyInputs: KeyInputs,
+      inputContracts: Set[ContractId],
+  )(implicit
+      contextualizedErrorLogger: ContextualizedErrorLogger
+  ): Either[Rejection, Unit] =
+    inputContracts.intersect(consumedContractsState) pipe {
+      case alreadyArchived if alreadyArchived.nonEmpty =>
+        Left(UnknownContracts(alreadyArchived)(errorFactories))
+      case _ =>
+        keyInputs
+          .foldLeft[Either[Rejection, Unit]](Right(())) {
+            case (Right(_), (key, LfTransaction.KeyCreate)) =>
+              keysState.get(key) match {
+                case None | Some((None, _)) => Right(())
+                case Some((Some(_), _)) => Left(DuplicateKey(key)(errorFactories))
+              }
+            case (Right(_), (key, LfTransaction.NegativeKeyLookup)) =>
+              keysState.get(key) match {
+                case None | Some((None, _)) => Right(())
+                case Some((Some(actual), _)) =>
+                  Left(InconsistentContractKey(None, Some(actual))(errorFactories))
+              }
+            case (Right(_), (key, LfTransaction.KeyActive(cid))) =>
+              keysState.get(key) match {
+                case None | Some((Some(`cid`), _)) => Right(())
```

Comment:
Should None be here or below ?

Reply:
On `None` we assume that the key has not been updated in the current sequencer state window. Since we trust the ledger state before the sequencer state window (i.e. the updates that we have already ingested) and nothing changed in the in-flight state, we assume no conflict.